File: src/SimDeck.Core/Protocol.cs
namespace SimDeck.Core;

/// <summary>
/// SimDeck wire protocol. Must stay byte-identical to protocol.py and
/// firmware/SimDeckClient/SimDeckClient.h.
///
/// Two planes, both UDP:
///   Control  JSON, low rate, module and hub both ways
///   Data     packed binary, high rate, hub to module
/// </summary>
public static class Protocol
{
    public const byte Version = 1;

    public const int CtrlPort   = 27500;   // hub listens
    public const int ModulePort = 27501;   // modules listen
    public const int HttpPort   = 27502;   // firmware images + local API

    public const byte DataMagic = 0x5A;

    public const byte FlagSimOk     = 0x01;
    public const byte FlagProfileOk = 0x02;

    /// <summary>
    /// magic, version, seq(2), count, flags, reserved(2).
    ///
    /// Eight bytes, not six. The two reserved bytes exist so the float
    /// payload starts 4-byte aligned, and so the header length is an
    /// obvious constant rather than something each implementation works
    /// out from a format string and gets wrong.
    /// </summary>
    public const int HeaderLen = 8;

    public const int MaxSlots = 16;

    public static byte[] EncodeFrame(ushort seq, byte flags, IReadOnlyList<float> values)
    {
        if (values.Count > MaxSlots)
            throw new ArgumentException($"too many slots: {values.Count}");

        var buf = new byte[HeaderLen + 4 * values.Count];
        buf[0] = DataMagic;
        buf[1] = Version;
        buf[2] = (byte)(seq & 0xFF);
        buf[3] = (byte)(seq >> 8);
        buf[4] = (byte)values.Count;
        buf[5] = flags;
        buf[6] = 0;
        buf[7] = 0;

        for (int i = 0; i < values.Count; i++)
        {
            // BitConverter is little endian on every platform we ship to,
            // but be explicit rather than assume it.
            var bytes = BitConverter.GetBytes(values[i]);
            if (!BitConverter.IsLittleEndian) Array.Reverse(bytes);
            Buffer.BlockCopy(bytes, 0, buf, HeaderLen + 4 * i, 4);
        }
        return buf;
    }

    public static bool TryDecodeFrame(ReadOnlySpan<byte> buf, out ushort seq,
                                      out byte flags, out float[] values)
    {
        seq = 0; flags = 0; values = Array.Empty<float>();
        if (buf.Length < HeaderLen) return false;
        if (buf[0] != DataMagic || buf[1] != Version) return false;

        seq = (ushort)(buf[2] | (buf[3] << 8));
        int count = buf[4];
        flags = buf[5];

        if (count > MaxSlots) return false;
        if (buf.Length < HeaderLen + 4 * count) return false;

        values = new float[count];
        for (int i = 0; i < count; i++)
            values[i] = BitConverter.ToSingle(buf.Slice(HeaderLen + 4 * i, 4));
        return true;
    }
}

File: bench-tools/protocol.py
"""SimDeck wire protocol.

Two planes, both UDP:

  Control plane  - JSON, low rate, module <-> hub
  Data plane     - packed binary, high rate, hub -> module

Keep this file in sync with firmware/SimDeckClient/SimDeckClient.h
"""

import struct

PROTO_VERSION = 1

CTRL_PORT = 27500     # hub listens here for HELLO / PING / EVENT
MODULE_PORT = 27501   # modules listen here for WELCOME and data frames

DATA_MAGIC = 0x5A

# flags byte in the data frame header
FLAG_SIM_OK = 0x01        # data source is alive
FLAG_PROFILE_OK = 0x02    # a matching aircraft profile is loaded

# magic, version, seq, count, flags, reserved(2)
#
# Eight bytes, not six. The reserved pair 4-byte aligns the float payload and
# makes the header length explicit rather than something each implementation
# derives from a format string. The firmware previously assumed 7 while this
# packed 6, so every frame was misread on real hardware.
_HEADER = struct.Struct("<BBHBBH")
HEADER_LEN = _HEADER.size  # 8

MAX_SLOTS = 64


def encode_frame(seq: int, flags: int, values) -> bytes:
    """Pack a data frame. values is an ordered list of floats (slot order).

    A value the hub could not resolve is sent as NaN; firmware treats NaN
    as 'no data' and holds the last good reading.
    """
    n = len(values)
    if n > MAX_SLOTS:
        raise ValueError("too many slots: %d" % n)
    head = _HEADER.pack(DATA_MAGIC, PROTO_VERSION, seq & 0xFFFF, n, flags, 0)
    return head + struct.pack("<%df" % n, *values)


def decode_frame(buf: bytes):
    """Returns (seq, flags, [floats]) or None if the packet is not ours."""
    if len(buf) < HEADER_LEN:
        return None
    magic, ver, seq, count, flags, _res = _HEADER.unpack_from(buf, 0)
    if magic != DATA_MAGIC or ver != PROTO_VERSION:
        return None
    if len(buf) < HEADER_LEN + 4 * count:
        return None
    vals = struct.unpack_from("<%df" % count, buf, HEADER_LEN)
    return seq, flags, list(vals)
